pyogrio/_err.pyx

```diff
@@ -248,3 +298,127 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+                    raise last
+            else:
+                raise CPLError(CE_Failure, err, "Unspecified OGR / GDAL error")
+
+        return err
+
+    cdef void *check_pointer(self, void *ptr) except NULL:
+        """Check the pointer returned by a GDAL/OGR function.
+
+        If `ptr` is `NULL`, an exception inheriting from CPLE_BaseError is
+        raised.
+        When a non-fatal GDAL/OGR error was captured in the error stack, the
+        exception raised will be customized appropriately. Otherwise, a
+        NullPointerError is raised.
+        """
+        if ptr == NULL:
+            stack = self.error_stack.get()
+            for error, cause in zip_longest(stack[::-1], stack[::-1][1:]):
+                if error is not None and cause is not None:
+                    error.__cause__ = cause
+
+            if stack:
+                last = stack.pop()
+                if last is not None:
+                    raise last
+            else:
+                raise NullPointerError(-1, -1, "NULL pointer error")
+
+        return ptr
+
+
+cdef void stacking_error_handler(
+    CPLErr err_class,
+    int err_no,
+    const char* err_msg
+) noexcept nogil:
+    """Custom CPL error handler that adds non-fatal errors to a stack.
+
+    All non-fatal errors (CE_Failure) are not printed to stderr (behaviour
+    of the default GDAL error handler), but they are converted to python
+    exceptions and added to a stack, so they can be dealt with afterwards.
+
+    Warnings are converted to Python warnings.
+    """
+    if err_class == CE_Fatal:
+        # If the error class is CE_Fatal, we want to have a message issued
+        # because the CPL support code does an abort() before any exception
+        # can be generated
+        CPLDefaultErrorHandler(err_class, err_no, err_msg)
+        return
+
+    elif err_class == CE_Failure:
+        # For Failures, add them to the error exception stack
+        with gil:
+            global _ERROR_STACK
+            stack = _ERROR_STACK.get()
+            stack.append(
+                exception_map.get(err_no, CPLE_BaseError)(
+                    err_class, err_no, clean_error_message(err_msg)
+                ),
+            )
+            _ERROR_STACK.set(stack)
+
+        return
+
+    elif err_class == CE_Warning:
+        with gil:
+            warnings.warn(clean_error_message(err_msg), RuntimeWarning)
+        return
+
+    # Fall back to the default handler for non-failure messages since
+    # they won't be translated into exceptions.
+    CPLDefaultErrorHandler(err_class, err_no, err_msg)
+
+
+@contextlib.contextmanager
+def capture_errors():
+    """A context manager that captures all GDAL non-fatal errors occuring.
+
+    It adds all errors to a single stack, so it assumes that no more than one
+    GDAL function is called.
+
+    Yields an ErrorHandler object that can be used to handle the errors
+    if any were captured.
+    """
+    CPLErrorReset()
+    global _ERROR_STACK
```

Comment:
What is the reason this needs to be a global variable? Because otherwise `stacking_error_handler` cannot access it? (being a `cdef` function, we would not be able to pass a python object as argument it?)

Reply:
Ah, and also of course the `stacking_error_handler` has a fixed signature and gets called from GDAL, so if we want to pass some custom data to that function, that's what the user data could be used for in `CPLPushErrorHandlerEx` (and retrieved with `CPLGetErrorHandlerUserData`).

I suppose we could use that to pass through the error stack, but at that point using a global variable might be simpler

Reply:
But, isn't the point of a context var that you don't need to use `global` for it to work? Does it still work if you just remove this line?

Reply:
I had a quick read, and indeed... I'd guess as well that the `global ...` statements shouldn't be there. I remover them, and the tests still pass without issues.